File: Problems/No.1654_Minimum_Jumps_to_Reach_Home/minimum_jumps_to_reach_home.py
class Solution(object):
    def minimumJumps(self, forbidden, a, b, x):
        """
        :type forbidden: List[int]
        :type a: int
        :type b: int
        :type x: int
        :rtype: int
        """

        # parameters
        jumps = -1
        possibleJumps = []

        # start computing
        possibleJumps.append([0, True])
        while possibleJumps:
            jumps += 1
            pP_size = len(possibleJumps)

            for i in range(pP_size):
                [currP, backC] = possibleJumps.pop(0)

                if (currP == x):
                    return jumps
                elif (currP in forbidden):
                    continue
                else:
                    
                    # backward (check if it backwarded once, also for the negative position)
                    if (backC and (0 <= currP-b)):
                        possibleJumps.append([currP-b, False])
                    
                    # forward (check if it is going to jump redundandly)
                    if (currP-b <= 2000):
                        possibleJumps.append([currP+a, True])
                
                forbidden.append(currP)

        return -1

File: Problems/No.1654_Minimum_Jumps_to_Reach_Home/minimum_jumps_to_reach_home.cpp
class Solution {
public:
    int minimumJumps(vector<int>& forbidden, int a, int b, int x) {
        
        // parameters
        int jumps = -1;
        std::unordered_set<int> forbiddenPositions(forbidden.begin(), forbidden.end());
        std::queue<std::pair<int, bool>> possibleJumps;

        // start computing
        possibleJumps.push({0, true});
        while (!possibleJumps.empty()) {

            jumps++;
            auto pP_size = possibleJumps.size();

            for (int i=0; i<pP_size; i++) {

                auto [currP, backC] = possibleJumps.front();
                possibleJumps.pop();

                if (currP == x) return jumps;
                else if (forbiddenPositions.find(currP) != forbiddenPositions.end()) continue;
                else {

                    // backward (check if it backwarded once, also for the negative position)
                    if (backC && (0 <= currP-b)) possibleJumps.push({currP-b, false});

                    // forward (check if it is going to jump redundandly)
                    if (currP-b <= 2000) possibleJumps.push({currP+a, true});

                }

                forbiddenPositions.insert(currP);
            
            }

        }

        return -1;

    }
};
